go_lib/hooks/generate_password/hook.go

```diff
@@ -17,53 +17,211 @@ limitations under the License.
+	if h.Secret.RawPasswordField != "" {
+		return string(secret.Data[h.Secret.RawPasswordField]), nil
 	}
 
-	generatedPass := pwgen.AlphaNum(20)
+	// Return field with basic auth.
+	return string(secret.Data[h.SecretField()]), nil
+}
 
-	input.ConfigValues.Set(fmt.Sprintf("%s.auth.password", moduleValuesPath), generatedPass)
+// Handle restores password from the configuration or from the Secret and
+// puts it to internal values.
+// It generates new password if no password found in the configuration and the
+// Secret or no externalAuthentication defined.
+func (h *Hook) Handle(input *go_hook.HookInput) error {
+	var externalAuthKey = h.ExternalAuthKey()
+	var passwordKey = h.PasswordKey()
+	var passwordInternalKey = h.PasswordInternalKey()
+
+	// Clear password from internal values if an external authentication is enabled.
+	if input.Values.Exists(externalAuthKey) {
+		input.Values.Remove(passwordInternalKey)
+		return nil
+	}
 
-	return nil
-}
+	// Try to set password from config values.
+	password, ok := input.ConfigValues.GetOk(passwordKey)
+	if ok {
+		input.Values.Set(passwordInternalKey, password.String())
+		return nil
+	}
 
-func generatePassword(moduleValuesPath string) func(input *go_hook.HookInput) error {
-	return func(input *go_hook.HookInput) error {
-		err := generatePasswordWithArgs(input, moduleValuesPath)
+	// Try to restore password from the Secret.
+	snap := input.Snapshots[secretBindingName]
+	if len(snap) > 0 {
+		secretField, ok := snap[0].(string)
+		if !ok {
+			return fmt.Errorf("problem getting field '%s' from Secret/%s: got %T, while string is expected", h.SecretField(), h.Secret.Name, snap[0])
+		}
+		storedPassword, err := h.extractPasswordFromBasicAuth(secretField)
 		if err != nil {
 			return err
 		}
+		input.Values.Set(passwordInternalKey, storedPassword)
 		return nil
 	}
+
+	// No config value, no Secret, generate new password.
+	newPassword := ""
+	if h.PasswordGeneratorFunc != nil {
+		newPassword = h.PasswordGeneratorFunc(input)
+	} else {
+		newPassword = DefaultGenerator()
+	}
+
+	input.Values.Set(passwordInternalKey, newPassword)
+	return nil
+}
+
+const (
+	externalAuthKeyTmpl     = "%s.auth.externalAuthentication"
+	passwordKeyTmpl         = "%s.auth.password"
+	passwordInternalKeyTmpl = "%s.internal.auth.password"
+)
+
+func (h *Hook) ExternalAuthKey() string {
+	key := h.Values.ExternalAuthKey
+	if key != "" {
+		return key
+	}
+	return fmt.Sprintf(externalAuthKeyTmpl, h.Values.ModuleKey)
+}
+
+func (h *Hook) PasswordKey() string {
+	key := h.Values.PasswordKey
+	if key != "" {
+		return key
+	}
+	return fmt.Sprintf(passwordKeyTmpl, h.Values.ModuleKey)
+}
+
+func (h *Hook) PasswordInternalKey() string {
+	key := h.Values.PasswordInternalKey
+	if key != "" {
+		return key
+	}
+	return fmt.Sprintf(passwordInternalKeyTmpl, h.Values.ModuleKey)
+}
+
+func (h *Hook) SecretField() string {
+	field := h.Secret.BasicAuthPlainField
+	if field == "" {
+		field = defaultBasicAuthPlainField
```

Comment:
Are there any cases to use diffirent fields?

Reply:
That field is to use hook with openvpn module, but turns out not working. Deleted.